tonic/models/vic/grid_params.py

```diff
@@ -661,64 +1218,172 @@ def grid_params(soil_dict, target_grid, snow_dict, veg_dict, veglib_dict,
+                    new[-1, :, :] = bare_vegparam[var]
+                else:
+                    new[:, :, :] = out_dicts['veg_dict'][var]
+                out_dicts['veg_dict'][var] \
+                    = np.ma.masked_values(new, FILLVALUE_F)
+
         # Distribute the veglib variables
         # 1st - the 1d vars
         #   double lib_overstory(veg_class) ;  --> (veg_class, nj, ni)
-        for var in ['overstory', 'rarc', 'rmin', 'wind_h', 'RGL', 'rad_atten',
-                    'rad_atten', 'wind_atten', 'trunk_ratio', 'snow_albedo']:
+        varnames = ['overstory', 'rarc', 'rmin', 'wind_h', 'RGL', 'rad_atten',
+                    'rad_atten', 'wind_atten', 'trunk_ratio']
+        if veglib_photo:
+            varnames.append('Ctype')
+            varnames.append('MaxCarboxRate')
+            varnames.append('MaxE_or_CO2Spec')
+            varnames.append('LUE')
+            varnames.append('Nscale')
+            varnames.append('Wnpp_inhib')
+            varnames.append('NPPfactor_sat')
+        for var in varnames:
             lib_var = 'lib_{0}'.format(var)
-            new = np.zeros((nveg_clases, ysize, xsize)) + FILLVALUE_F
-            new[:-1, yi, xi] = veglib_dict[lib_var][:, np.newaxis]
-            new[-1, yi, xi] = 0
+            if var in ['Ctype', 'Nscale']:
+                fill_val = FILLVALUE_I
+                new = np.full((nveg_classes, ysize, xsize), fill_val,
+                              dtype=np.int)
+            else:
+                fill_val = FILLVALUE_F
+                new = np.full((nveg_classes, ysize, xsize), fill_val)
+            if extra_class:
+                new[:-1, yi, xi] = veglib_dict[lib_var][:, np.newaxis]
+                new[-1, yi, xi] = bare_vegparam[var]
+            else:
+                new[:, yi, xi] = veglib_dict[lib_var][:, np.newaxis]
             new[:, ymask, xmask] = fill_val
-            out_dicts['veg_dict'][var] = np.ma.masked_values(new, FILLVALUE_F)
+            out_dicts['veg_dict'][var] = np.ma.masked_values(new, fill_val)
 
         # 2nd - the 2d vars
-        for var in ['albedo', 'veg_rough', 'displacement']:
+        varnames = ['veg_rough', 'displacement']
+        if alb_src == 'FROM_VEGLIB':
+            varnames = ['albedo'] + varnames
+        if veglib_fcan and fcan_src == 'FROM_VEGLIB':
+            varnames = ['fcanopy'] + varnames
+        if lai_src == 'FROM_VEGLIB':
+            varnames = ['LAI'] + varnames
+        for var in varnames:
             lib_var = 'lib_{0}'.format(var)
-            shape = (nveg_clases, veglib_dict[lib_var].shape[1], ysize, xsize)
-            new = np.zeros(shape) + FILLVALUE_F
-            new[:-1, :, yi, xi] = veglib_dict[lib_var][:, :, np.newaxis]
-            new[-1, :, yi, xi] = 0
+            shape = (nveg_classes, veglib_dict[lib_var].shape[1],
+                     ysize, xsize)
+            new = np.full(shape, FILLVALUE_F)
+            if extra_class:
```

Comment:
since `fill_val` will have already been assigned to the integer or float fillvalue